src/main/java/org/myrobotlab/service/UltrasonicSensor.java

```diff
@@ -113,11 +121,51 @@ public void attach(UltrasonicSensorController controller, Integer trigPin, Integ
 
     // call other service's attach
     controller.attach(this, trigPin, echoPin);
+    isAttached = true;
+    broadcastState();
+  }
 
+  public void clear() {
+    pingCount = 0;
+    lastRange = null;
+    min = null;
+    max = null;
+    broadcastState();
   }
 
-  private boolean isAttached(UltrasonicSensorController controller) {
-    return this.controller == controller;
+  // TODO - this could be Java 8 default interface implementation
+  @Override
+  public void detach(String controllerName) {
+    if (controller == null || !controllerName.equals(controller.getName())) {
+      return;

```

Comment:
this should probably set isAttached=false here...  consider the scenario where you attach the ultrasonic sensor.. then someone via python, sets the controller =null..   you would never be able to detach properly..

Reply:
agree - fixed